apps/opik-frontend/src/components/pages/PlaygroundPage/PlaygroundOutputs/PlaygroundOutputScores/PlaygroundOutputScores.tsx

```diff
@@ -0,0 +1,356 @@
+        const receivedScores = data?.feedback_scores ?? [];
+        const expectedNames = expectedScoreNamesRef.current;
+
+        // If specific rules were selected, wait for all expected scores
+        if (expectedNames.size > 0) {
+          const receivedNames = new Set(receivedScores.map((s) => s.name));
+          const allExpectedReceived = [...expectedNames].every((name) =>
+            receivedNames.has(name),
+          );
+          if (allExpectedReceived) {
+            return false;
+          }
+        } else {
+          // For "all" selection or legacy data, stop as soon as we have any scores
+          if (receivedScores.length > 0) {
+            return false;
+          }
+        }
+
+        return REFETCH_INTERVAL;
+      },
+    },
+  );
+
+  const feedbackScores: TraceFeedbackScore[] = useMemo(() => {
+    return trace?.feedback_scores ?? [];
+  }, [trace?.feedback_scores]);
+
+  // Create a map from score name to score for quick lookup
+  const scoresByName = useMemo(() => {
+    const map: Record<string, TraceFeedbackScore> = {};
+    for (const score of feedbackScores) {
+      map[score.name] = score;
+    }
+    return map;
+  }, [feedbackScores]);
+
+  // Build list of expected metrics with their display info, sorted by score name
+  // Note: We deduplicate by score name because TraceFeedbackScore doesn't include rule_id,
+  // so if multiple rules produce the same metric name, we can only show one entry
+  const expectedMetrics = useMemo(() => {
+    const metrics: Array<{
+      scoreName: string;
+      score: TraceFeedbackScore | null;
+    }> = [];
+    const seenScoreNames = new Set<string>();
+
+    // If selectedRuleIds is null/undefined (meaning "all" was selected or legacy data),
+    // only show scores that actually exist in the trace - no loading spinners
+    if (selectingAllRules) {
+      for (const score of feedbackScores) {
+        if (!seenScoreNames.has(score.name)) {
+          seenScoreNames.add(score.name);
+          metrics.push({
+            scoreName: score.name,
+            score,
+          });
+        }
+      }
+    } else {
+      // Specific rules were selected - need rules to load to know expected scores
+      if (!rulesLoaded) return [];
+
+      for (const rule of selectedRules) {
+        const scoreNames = getScoreNamesFromRule(rule);
+        for (const scoreName of scoreNames) {
+          // Deduplicate: if multiple rules produce the same metric name, show only once
+          if (!seenScoreNames.has(scoreName)) {
+            seenScoreNames.add(scoreName);
+            metrics.push({
+              scoreName,
+              score: scoresByName[scoreName] || null,
+            });
+          }
+        }
+      }
+    }
+
+    // Sort by score name alphabetically
+    return metrics.sort((a, b) => a.scoreName.localeCompare(b.scoreName));
+  }, [
+    rulesLoaded,
+    selectedRules,
+    scoresByName,
+    selectingAllRules,
+    feedbackScores,
+  ]);
```

Comment:
```suggestion
const expectedMetrics = useMemo(() => {
  if (!selectingAllRules && !rulesLoaded) return [];

  const sourceNames = selectingAllRules 
    ? feedbackScores.map(s => s.name)
    : Array.from(expectedScoreNames);

  const metrics = Array.from(new Set(sourceNames)).map((name) => ({
    scoreName: name,
    score: scoresByName[name] ?? null,
  }));

  return metrics.sort((a, b) => a.scoreName.localeCompare(b.scoreName));
}, [
  rulesLoaded,
  expectedScoreNames,
  scoresByName,
  selectingAllRules,
  feedbackScores,
]);
```

By using expectedScoreNames - which is a Set, we can remove the nested loops and the manual "seen" tracking

This logic should be checked/tested